contracts/AirDrop.sol

```diff
@@ -0,0 +1,186 @@
+pragma solidity 0.6.4;
+
+import "./interface/IBEP20.sol";
+import "./interface/ITokenHub.sol";
+import "./interface/IAirDrop.sol";
+import "./interface/IParamSubscriber.sol";
+import "./lib/SafeMath.sol";
+import "./lib/BytesToTypes.sol";
+import "./System.sol";
+import "./MerkleProof.sol";
+
+contract AirDrop is IAirDrop, IParamSubscriber, System {
+    using SafeMath for uint256;
+
+    string public constant sourceChainID = "Binance-Chain-Ganges";
+    address public approvalAddress = 0xaAaAaAaaAaAaAaaAaAAAAAAAAaaaAaAaAaaAaaAa;
+    bytes32 public override merkleRoot = 0x0000000000000000000000000000000000000000000000000000000000000000;
+    bool public merkleRootAlreadyInit = false;
+
+    // claimedMap is used to record the claimed token.
+    mapping(bytes32 => bool) private claimedMap;
+
+    function isClaimed(bytes32 node) public view override returns (bool) {
+        return claimedMap[node];
+    }
+
+    function claim(
+        bytes32 tokenSymbol, uint256 amount,
+        bytes calldata ownerPubKey, bytes calldata ownerSignature, bytes calldata approvalSignature,
+        bytes32[] calldata merkleProof) external override {
+        // Recover the owner address and check signature.
+        bytes memory ownerAddr = _verifyTMSignature(ownerPubKey, ownerSignature, _tmSignatureHash(tokenSymbol, amount, msg.sender));
+        // Generate the leaf node of merkle tree.
+        bytes32 node = keccak256(abi.encodePacked(ownerAddr, tokenSymbol, amount));
+    
+        // Check if the token is claimed.
+        require(isClaimed(node), "AlreadyClaimed");
+
+        address contractAddr = address(0x00);
+        // Check if the token is exist.
+        if (tokenSymbol != "BNB") {
+            contractAddr = _checkTokenContractExist(tokenSymbol);
+        }
+        
+        // Verify the approval signature.
+        _verifySignature(msg.sender, ownerSignature, approvalSignature, node, merkleProof);
+    
+        // Verify the merkle proof.
+        require(MerkleProof.verify(merkleProof, merkleRoot, node), "InvalidProof");
+    
+        // Unlock the token from TokenHub.
+        ITokenHub(TOKEN_HUB_ADDR).unlock(contractAddr, msg.sender, amount);
+
+        // Mark it claimed and send the token.
+        claimedMap[node] = true;
+
+        emit Claimed(tokenSymbol, msg.sender, amount);
+    }
+
+    function _verifySignature(address account, bytes memory ownerSignature, bytes memory approvalSignature, bytes32 leafHash, bytes32[] memory merkleProof) private view {
+        // Ensure the account is not the zero address
+        require(account != address(0), "InvalidSignature");
+
+        // Ensure the signature length is correct
+        require(approvalSignature.length == 65, "InvalidSignature");
+        bytes32 r;
+        bytes32 s;
+        uint8 v;
+        assembly {
```

Comment:
What is the reference for these code pieces?

Reply:
it's reference to https://github.com/OpenZeppelin/openzeppelin-contracts/blob/0950532d9a89ac652dbbb2aed56678bdfd624c21/contracts/utils/cryptography/ECDSA.sol#L56

Reply:
after upgrade to 0.8.17, I use the original openzeppelin lib